packages/framework/src/Model/Article/ArticleFriendlyUrlDataProvider.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Model\Article;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Query\Expr\Join;
+use Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig;
+use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\CompilerPass\FriendlyUrlDataProviderInterface;
+use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrl;
+use Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData;
+
+class ArticleFriendlyUrlDataProvider implements FriendlyUrlDataProviderInterface
+{
+    private const ROUTE_NAME = 'front_article_detail';
+    /**
+     * @var \Doctrine\ORM\EntityManagerInterface
+     */
+    private $em;
+
+    public function __construct(
+        EntityManagerInterface $em
+    ) {
+        $this->em = $em;
+    }
+
+    /**
+     * Returns data required for friendly url generator
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig $domainConfig
+     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData[]
+     */
+    public function getFriendlyUrlData(DomainConfig $domainConfig): array
+    {
+        $queryBuilder = $this->em->createQueryBuilder()
+            ->select('a.id, a.name')
+            ->distinct()
+            ->from(Article::class, 'a')
+            ->leftJoin(FriendlyUrl::class, 'f', Join::WITH, 'a.id = f.entityId AND f.routeName = :routeName AND f.domainId = a.domainId')
+            ->setParameter('routeName', self::ROUTE_NAME)
+            ->where('f.entityId IS NULL AND a.domainId = :domainId')
+            ->setParameter('domainId', $domainConfig->getId());
+        $scalarData = $queryBuilder->getQuery()->getScalarResult();
+
+        $friendlyUrlsData = [];
+
+        foreach ($scalarData as $data) {
```

Comment:
As we discussed, this part could may be in new `FriendlyUrlDataFactory`...we can talk about the particular solution personally - maybe it could look like the deleted `FriendlyUrlToGenerateRepository::createFriendlyUrlsData()` method